File: apps/app-openapi-rest/go/repository/data_sequence_repository.go
package repository

import (
	"context"
	"encoding/base64"
	"strconv"

	"github.com/go-redis/redis/v8"
)

type DataSequence struct {
	ID             string
	Data           []byte
	DataOffsetStart int
	DataOffsetEnd   int
	ResequenceCount int
	OwnerID        string
	Timestamp      string
	PurgeExpiry    string
}

type DataSequenceRepository struct {
	redisClient *redis.Client
	ownerRepo   *OwnerRepository
}

func NewDataSequenceRepository(redisClient *redis.Client, ownerRepo *OwnerRepository) *DataSequenceRepository {
	return &DataSequenceRepository{redisClient: redisClient, ownerRepo: ownerRepo}
}

func (r *DataSequenceRepository) CreateDataSequence(ctx context.Context, ds DataSequence) error {
	encodedData := base64.StdEncoding.EncodeToString(ds.Data)

	err := r.redisClient.HSet(ctx, "data_sequence:"+ds.ID, map[string]interface{}{
		"ID":              ds.ID,
		"Data":            encodedData,
		"DataOffsetStart": ds.DataOffsetStart,
		"DataOffsetEnd":   ds.DataOffsetEnd,
		"ResequenceCount": ds.ResequenceCount,
		"OwnerID":         ds.OwnerID,
		"Timestamp":       ds.Timestamp,
		"PurgeExpiry":     ds.PurgeExpiry,
	}).Err()
	if err != nil {
		return err
	}

	return r.ownerRepo.AddDataSequenceToOwner(ctx, ds.OwnerID, ds.ID)
}

func (r *DataSequenceRepository) GetDataSequenceByID(ctx context.Context, id string) (*DataSequence, error) {
	data, err := r.redisClient.HGetAll(ctx, "data_sequence:"+id).Result()
	if err != nil || len(data) == 0 {
		return nil, err
	}

	decodedData, _ := base64.StdEncoding.DecodeString(data["Data"])

	return &DataSequence{
		ID:              data["ID"],
		Data:            decodedData,
		DataOffsetStart: r.intFromString(data["DataOffsetStart"]),
		DataOffsetEnd:   r.intFromString(data["DataOffsetEnd"]),
		ResequenceCount: r.intFromString(data["ResequenceCount"]),
		OwnerID:         data["OwnerID"],
		Timestamp:       data["Timestamp"],
		PurgeExpiry:     data["PurgeExpiry"],
	}, nil
}

func (r *DataSequenceRepository) GetAllDataSequences(ctx context.Context) ([]DataSequence, error) {
	keys, err := r.redisClient.Keys(ctx, "data_sequence:*").Result()
	if err != nil {
		return nil, err
	}

	var sequences []DataSequence
	for _, key := range keys {
		ds, err := r.GetDataSequenceByID(ctx, key[14:]) // remove "data_sequence:" prefix
		if err == nil {
			sequences = append(sequences, *ds)
		}
	}

	return sequences, nil
}

func (r *DataSequenceRepository) intFromString(s string) int {
	// Utility function to convert a string to an int, assuming valid input
	i, _ := strconv.Atoi(s)
	return i
}

File: apps/app-openapi-rest/nodejs/repositories/dataSequenceRepository.mjs
import { createClient } from 'redis';

class DataSequenceRepository {
  constructor() {
    this.redisClient = createClient({
      url: `redis://${process.env.REDIS_HOST}:${process.env.REDIS_PORT}`,
    });
    this.redisClient.on('error', (err) => console.error('Redis Client Error', err));
    this.redisClient.connect();
  }

  async createDataSequence(dataSequence) {
    const { ID, Data, DataOffsetStart, DataOffsetEnd, ResequenceCount, OwnerID, Timestamp, PurgeExpiry } = dataSequence;

    // Store the DataSequence in Redis as a hash
    await this.redisClient.hSet(`data_sequence:${ID}`, {
      ID,
      Data: Buffer.from(Data).toString('base64'), // Encode the byte array as a base64 string
      DataOffsetStart,
      DataOffsetEnd,
      ResequenceCount,
      OwnerID,
      Timestamp,
      PurgeExpiry,
    });

    // Add the DataSequence ID to the owner's list of sequences
    await this.ownerRepo.addDataSequenceToOwner(OwnerID, ID);
  }

  async getDataSequenceById(id) {
    const dataSequence = await this.redisClient.hGetAll(`data_sequence:${id}`);

    if (dataSequence && Object.keys(dataSequence).length > 0) {
      dataSequence.Data = Buffer.from(dataSequence.Data, 'base64'); // Decode the base64 Data back to binary
      return dataSequence;
    }
    return null;
  }

  async getAllDataSequences() {
    const keys = await this.redisClient.keys('data_sequence:*');
    const sequences = [];

    for (const key of keys) {
      const dataSequence = await this.redisClient.hGetAll(key);
      if (dataSequence && Object.keys(dataSequence).length > 0) {
        dataSequence.Data = Buffer.from(dataSequence.Data, 'base64'); // Decode the base64 Data back to binary
        sequences.push(dataSequence);
      }
    }

    return sequences;
  }
}

export default DataSequenceRepository;
